modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -74,48 +172,111 @@ object GlobalL0Service {
+            } yield pulled.asRight[LatestSnapshotTuple]
           }
         }.handleErrorWith { e =>
-          logger.warn(e)(s"Failure pulling global snapshots!") >>
+          logger.warn(e)("Failure pulling global snapshots") >>
             Applicative[F].pure(
-              List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)]
+              List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[LatestSnapshotTuple]
             )
         }
+
+      private def pullSnapshots(
+        l0Peer: L0Peer,
+        nextOrdinal: SnapshotOrdinal,
+        lastOrdinal: SnapshotOrdinal
+      ): F[List[Hashed[GlobalIncrementalSnapshot]]] = {
+        val lastOrdinalCap = lastOrdinal.value.value
+          .min(singlePullLimit.map(nextOrdinal.value.value + _.value).getOrElse(lastOrdinal.value.value))
+
+        val ordinals = (nextOrdinal.value.value to lastOrdinalCap).toList
+          .map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o)))
+
+        (ordinals, List.empty[Hashed[GlobalIncrementalSnapshot]]).tailRecM {
+          case (ordinal :: nextOrdinals, snapshots) =>
+            l0GlobalSnapshotClient
+              .get(ordinal)(l0Peer)
+              .flatMap(_.toHashedWithSignatureCheck.flatMap(_.liftTo[F]))
+              .map(s => (nextOrdinals, snapshots :+ s).asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+              .handleErrorWith { e =>
+                logger
+                  .warn(e)(s"Failure pulling snapshot with ordinal=${ordinal.show}")
+                  .map(_ => snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
+              }
+
+          case (Nil, snapshots) =>
+            Applicative[F].pure(snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
+        }
+      }
+
+      private case class MajoritySnapshotData(ordinal: SnapshotOrdinal, hash: Hash)
+
+      private def getLatestMajoritySnapshotData(peerIds: NonEmptyList[PeerId]): F[MajoritySnapshotData] = {
+        val maybeData =
+          for {
+            peers <- OptionT(getL0Peers(peerIds))
+            majorityOrdinal <- OptionT(getMajorityOrdinal(peers))
+            majorityHash <- OptionT(getMajorityHash(peers, majorityOrdinal))
+          } yield MajoritySnapshotData(majorityOrdinal, majorityHash)
+
+        maybeData.getOrElseF(NoMajoritySnapshotData.raiseError[F, MajoritySnapshotData])
+      }
+
+      private def getMajorityHash(peers: NonEmptyList[L0Peer], ordinal: SnapshotOrdinal): F[Option[Hash]] =
+        peers.toList
+          .parTraverseN(numConcurrentQueries)(p =>
+            attemptOption(l0GlobalSnapshotClient.getHash(ordinal).run(p), s"Unable to obtain hash for majority peer ${p.show}")
+          )
+          .map(_.flatMap(_.flatten))
+          .flatTap(hashes => logger.debug(s"** Majority Hashes ${hashes.show}"))
+          .map(selectMajorityHash(_, peers.size))
```

Comment:
I think that the `totalPeers` value is a constant and it is equal to count of priority peers declared by the user through configuration and not count of peers passed to this function. I would not assume that all of them will always be returned from the cluster storage - I would think about a scenario where a network restarts for example. I would also look into the GL0 peers discovery process to check what's gonna happen during GL0 restart and how will that influence the cluster storage to be sure that eventually the majority will be picked in every foreseeable scenario (or rightfully get stuck if majority peers don't align as they should)